WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadService.java

```diff
@@ -652,39 +711,44 @@ public void onMediaUploaded(OnMediaUploaded event) {
-                synchronized (sPostsWithPendingMedia) {
-                    Iterator<UploadingPost> iterator = sPostsWithPendingMedia.iterator();
-                    while (iterator.hasNext()) {
-                        UploadingPost uploadingPost = iterator.next();
-                        if (!UploadService.hasPendingOrInProgressMediaUploadsForPost(uploadingPost.postModel)) {
-                            if (uploadingPost.isCancelled) {
-                                // Finished all media uploads for a post upload that was cancelled (probably because
-                                // it was re-opened in the editor) - we can just remove it from the list now
-                                iterator.remove();
-                            } else {
-                                // Fetch latest version of the post, in case it has been modified elsewhere
-                                PostModel latestPost = mPostStore.getPostByLocalPostId(uploadingPost.postModel.getId());
-
-                                // Replace local with remote media in the post content
-                                PostModel updatedPost = updatePostWithCurrentlyCompletedUploads(latestPost);
-                                // also do the same now with failed uploads
-                                updatedPost = updatePostWithCurrentlyFailedUploads(updatedPost);
-                                // finally, save the PostModel
-                                mDispatcher.dispatch(PostActionBuilder.newUpdatePostAction(updatedPost));
-
-                                // TODO Should do some extra validation here
-                                // e.g. what if the post has local media URLs but no pending media uploads?
-                                iterator.remove();
-                                mPostUploadHandler.upload(updatedPost);
-                            }
-                        }
+                cleanUp();
+            }
+            stopServiceIfUploadsComplete();
+        }
+    }
+
+    private void cleanUp() {
+        synchronized (sPostsWithPendingMedia) {
+            Iterator<UploadingPost> iterator = sPostsWithPendingMedia.iterator();
+            while (iterator.hasNext()) {
+                UploadingPost uploadingPost = iterator.next();
+                if (!UploadService.hasPendingOrInProgressMediaUploadsForPost(uploadingPost.postModel)) {
+                    if (uploadingPost.isCancelled) {
+                        // Finished all media uploads for a post upload that was cancelled (probably because
+                        // it was re-opened in the editor) - we can just remove it from the list now
+                        iterator.remove();
+                    } else {
+                        // Fetch latest version of the post, in case it has been modified elsewhere
+                        PostModel latestPost = mPostStore.getPostByLocalPostId(uploadingPost.postModel.getId());
```

Comment:
Another good observation - this piece of code actually was written for the case when we're uploading things in the background. For the case where the user is actually editing the post, the Editor always has the last say on this, so chances of losing data are minimum (i.e. they can always stop the app while in the editor and before the autosave expires, that would be the only window for data loss in this case, unless they insert some media in which case we save things again as per https://github.com/wordpress-mobile/WordPress-Android/pull/6499).